quickwit/quickwit-query/src/query_ast/tantivy_query_ast.rs

```diff
@@ -144,12 +180,40 @@ impl TantivyBoolQuery {
                 }
             }
         }
-        self.must
-            .retain(|ast| ast.const_predicate() != Some(MatchAllOrNone::MatchAll));
-        self.filter
-            .retain(|ast| ast.const_predicate() != Some(MatchAllOrNone::MatchAll));
-        self.must_not
-            .retain(|child| child.const_predicate() != Some(MatchAllOrNone::MatchNone));
+        if self.should.is_empty()
+            && self.must.is_empty()
+            && self.filter.is_empty()
+            && self.must_not.is_empty()
+        {
+            // This is just a convention mimicking Elastic/Commonsearch's behavior.
+            return TantivyQueryAst::match_all();
+        }
+        remove_with_guard(&mut self.must, MatchAllOrNone::MatchAll, true);
+        let mut has_no_positive_ast_so_far = self.must.is_empty();
+        remove_with_guard(
+            &mut self.filter,
+            MatchAllOrNone::MatchAll,
+            has_no_positive_ast_so_far,
+        );
+        has_no_positive_ast_so_far &= self.filter.is_empty();
+        remove_with_guard(
+            &mut self.should,
+            MatchAllOrNone::MatchNone,
+            has_no_positive_ast_so_far,
+        );
+        has_no_positive_ast_so_far &= self.should.is_empty();
+        remove_with_guard(
+            &mut self.must_not,
+            MatchAllOrNone::MatchNone,
+            has_no_positive_ast_so_far,
+        );
+        if self.must.is_empty()
+            && self.should.is_empty()
+            && self.must_not.is_empty()
+            && self.filter.is_empty()
+        {
+            return TantivyQueryAst::match_none();
+        }
```

Comment:
I have a hard time wrapping my head around this bit.
1st I don't think I understand why the same condition above should `match_all`, but this one, after more filtering, should `match_none`.
2nd I'm not convinced the condition can ever evaluate to `true`. If any of the 4 lists is not empty before, due to the cascading `stop_before_empty = has_no_positive_ast_so_far`, then at least one of the list shouldn't be empty afterward?

Reply:
you are correct!